ark/utils/data_utils.py

```diff
@@ -46,7 +47,131 @@ def label_cells_by_cluster(fovs, all_data, label_maps, fov_col=settings.FOV_ID,
+        raise FileNotFoundError(
+            "consensus_dir %s does not exist in base_dir %s" % (cell_consensus_name, base_dir)
+        )
+
+    # load the consensus data in
+    cell_consensus_data = feather.read_dataframe(os.path.join(base_dir, cell_consensus_name))
+
+    # define the files for whole cell and nuclear
+    whole_cell_files = [fov + '_feature_0.tif' for fov in fovs]
+    # nuclear_files = [fov + '_feature_1.tif' for fov in fovs]
+
+    # load the segmentation labels in
+    label_maps = load_utils.load_imgs_from_dir(data_dir=seg_dir,
+                                               files=whole_cell_files,
+                                               xr_dim_name='compartments',
+                                               xr_channel_names=['whole_cell'],
+                                               trim_suffix='_feature_0',
+                                               force_ints=True)
+
+    # use label_cells_by_cluster to create cell masks
+    img_data = label_cells_by_cluster(
+        fovs, cell_consensus_data, label_maps, fov_col='fov',
+        cell_label_column='segmentation_label', cluster_column=cluster_col
+    )
+
+    return img_data
+
+
+def generate_pixel_cluster_mask(fovs, base_dir, seg_dir, pixel_consensus_dir,
+                                cluster_col='cluster'):
+    """For each fov, create a mask labeling each pixel with their SOM or meta cluster label
+
+    Args:
+        fovs (list):
+            List of fovs to relabel
+        base_dir (str):
+            The path to the data directory
+        seg_dir (str):
+            The path to the segmentation data
+        pixel_consensus_dir (str):
+            The path to the data with both pixel SOM and meta cluster assignments
+        cluster_col (str):
+            Whether to assign SOM or meta clusters, needs to be 'cluster' or 'hCluster_cap'
+
+    Returns:
+        xarray.DataArray:
+            The labeled images (dims: ["fovs", "rows", "cols"])
+    """
+
+    # path checking
+    if not os.path.exists(seg_dir):
+        raise FileNotFoundError("seg_dir %s does not exist")
+
+    if not os.path.exists(os.path.join(base_dir, pixel_consensus_dir)):
+        raise FileNotFoundError(
+            "consensus_dir %s does not exist in base_dir %s" % (pixel_consensus_dir, base_dir)
+        )
+
+    # verify all the fovs are valid
+    verify_in_list(
+        provided_fov_files=[fov + '.feather' for fov in fovs],
+        consensus_fov_files=os.listdir(os.path.join(base_dir, pixel_consensus_dir))
+    )
+
+    # define a list to hold the overlays for each fov
+    img_data = []
+
+    for fov in fovs:
+        # read the pixel data for the fov
+        fov_data = feather.read_dataframe(
+            os.path.join(base_dir, pixel_consensus_dir, fov + '.feather')
+        )
+
+        # read the segmentation mask to determine size of pixel cluster mask
+        seg_mask = io.imread(os.path.join(seg_dir, fov + '_feature_0.tif'))
```

Comment:
Is the reason we can't use the dimensions of the feather file due to zero pixels being excluded?

Reply:
Yes, because of this we cannot assume that the pixel data for a fov will have the same dimensions as the segmentation mask.